### DISCUSSION TITLE
Code katas for ivy, with solutions

### DISCUSSION THREAD

**Discussion by arl** [robpike/ivy#147] [General]
I'd be great if we could share code katas that are fun to solve with ivy, with the solutions possibly.

I'm gonna start with the obvious:

 - [Russ Cox solves most of advent of code 2021 with ivy](https://youtube.com/playlist?list=PLrwpzH1_9ufMLOB6BAdzO08Qx-9jHGfGg&feature=share8)

I'll update with a list of problems I had compiled for a work meetup of apl-like languages. In the meantime, please feel free to complete.

**Comment by fzipp** [comment]
Not a kata, but a port of [this Fast Fourier Transform implementation](https://aplwiki.com/wiki/Fast_Fourier_transform#Simple_recursive_implementation):

```
op fft A =
	1 > (K=(N=count A)/2): A
	0 != (2 log N) mod 1: 'Length of the input vector must be a power of 2'
	odd  = fft (N rho 1 0) sel A
	even = fft (N rho 0 1) sel A
	T = even * ** 0j-2 * pi*(-1+iota K)/N
	(odd+T),odd-T

op ifft A = (conj fft conj A)/count A
```

The 2D FFT implementation:

```
op fft2d A =
	or/ 0 != (2 log rho A) mod 1: 'Matrix dimensions must be powers of 2'
	transp mix fft@ transp mix fft@ A

op ifft2d A = (conj fft2d conj A)/count flatten A
```

**Reply by tcolgate** [reply]
That fft is pretty awesome!

**Comment by robpike** [comment]
On line 2 of fft, we know that N is an integer, so a faster way to test it's a power of 2 is to see if n&n-1 is zero. I believe this is easier in ivy than in APL.

```
op isPow2 N = 0 == N & N-1
```

**Comment by robpike** [comment]
Yesterday I tried fizzbuzz. Here's the best I found, but I suspect there's a nicer solution to be found.

```
op fizzbuzz n = ((0==n mod 3) sel "Fizz"), (0==n mod 5) sel "Buzz"
```

**Comment by ghost** [comment]
Not really a kata, but I wrote ivy-solvers for all the puzzles of Advent of Code 2024. I uploaded the solvers to [Github](https://github.com/skias-onar/aoc2024). There is no doubt that most of them could be largely improved!

`ivy` is quite useable as it stands, and I missed only a few primitives, for efficiency and/or comfort:

- an element-wise apply (already discussed here): I would suggest `a @f@ b` where the arguments have the same dimension. The current `@f@` is, I think, the same as `o.f`
- a power operator, something like `b f***3 a` and also `f***=== a` to get a function fixed-point
- a slice window option for reduction
- a unique mask function. As a side note, I noticed that the unique primitive does not give the elements in the *natural* order.
- and finally, a key operator: this one may not be easy to implement

Many thanks for `ivy`! It was really fun to play with!

**Comment by fzipp** [comment]
I've been experimenting with color escape sequences in ANSI terminals:

```
# Wrap a value in an ANSI 24-bit color escape sequence. R G B each in 0..255.
op (r g b) rgb VALUE =
  ,/ '\033[38;2;' (text r) ';' (text g) ';' (text b) 'm' ('s' text VALUE) '\033[0m'

# GREY should be in [0.0, 1.0].
op GREY grey VALUE =
  (floor 255 * 3 rho GREY) rgb VALUE

# Display a matrix of values in [0.0, 1.0] as a grey scale block image for ANSI terminals.
op greyimg M =
  ,/ (,/ M @@grey '█') @, '\n'

# Radial wave pattern as an NxN matrix.
# k = cycles per unit radius
op N radwave (k phase) = 
  x = ((iota N) - N / 2) / N
  y = x
  R = sqrt (x ** 2) o.+ y ** 2
  1/2 + 1/2 * sin (2 * pi) * k * R + phase

greyimg 30 radwave (10 0)
```

<img width="470" height="953" alt="ivy-ansidemo" src="https://github.com/user-attachments/assets/e8d78c2d-8c2d-43ef-84a8-f6c176a51f0b" />
